### DISCUSSION TITLE
Suggestion for Feature

### DISCUSSION THREAD

**Discussion by fxmb** [medusajs/b2b-starter-medusa#7] [Ideas]
We are actually planning to build a B2B customisation ourselves. I just wanted to check if you are planning to implement this here already? Maybe it can also just serve as an inspiration:

1. We will have B2B customers who have the opportunity to "register their company" domain in the storefront so that every following signup with that domain would be automatically added to that company. Companies don't have to do that of course. 
2. In the user account their would be "Team" button where admins can manage, invite, remove team members, see their orders
3. All team members have access to shared addresses and maybe even payment methods.
4. We could assign prices just fie this particular company from admin or via the admin api.
5. We could issue vouchers just for the use of one specific company. 

This would require some store apis I guess. Is this something you guys are planning?

**Comment by shiloh-expanse** [comment]
I was starting a project to build very similar functionality! In my case, I needed the Team "admin" to be able to upload a list of the users because they don't all have the same domain for the emails. Basically the same core functionality though.

+1 on this! 

Shiloh Pettyjohn

**Comment by fxmb** [comment]
I am not actively working on this right now. It is more of an idea what the B2B use case should incorporate.

**Comment by VariableVic** [comment]
Hey guys, 

@fxmb thanks for the input, good stuff!

1. I've thought about this, and I really like the idea. It's not a priority but I'm definitely open to adding this later.
2. This is planned
3. Good idea, might implement to some extent.
4. This is planned
5. I think this should work out of the box by adding a company_id rule to a promotion. But I'll have to test to make sure.

**Comment by fxmb** [comment]
Thanks for getting back. Especially 2 should be rather straight forward (would be "just" some store endpoints acting on the data model). I have some concerns about the team member <> cart relationship:

1. We want to introduce sharable carts (e.g. Ben from Marketing puts in item A, then Jen from Engineering puts in item B for her project, they make one order overall). At the same time "private" carts should be possible. This would weaken the 1:1 relationship between carts and customers. I feel it should be cart => company => customers where one cart can only be assigned to one company id but n customer ids. Its definitely not super trivial I believe
2. There should be different levels of RBAC. But I think you implemented this? 
3. (More of a side note: We will introduce Comment functions on lime item level soon). So e.g. customer XY adds a comment for themselves on a cart item that they need internally. E.g. an internal parts number or something.

PS Thanks for moving into discussion. Couldnt find that option

**Reply by VariableVic** [reply]
1. I'm actually working on this today. The plan is to link carts and orders to the company. I want to introduce shareable carts later on, my thinking was that every employee can access carts belonging to their company, but only if someone shares the unique link with them. So there wont be a list of all open carts in the UI for example. That way, employees can decide to keep the cart private as long as they don't share the unique link with colleagues.
2. Not yet, but will be implemented. Employees can have an `admin` role to manage other employees and company details.
3. Nice, are you using the metadata object for this?

Discussions were disabled before, that's why you couldn't find them 👍🏻

**Reply by fxmb** [reply]
Thats awesome! 
1. yes that sounds exactly what would be required
2. OK
3. We try to not use metadata too much. It feels like its convoluting the database / types and like a hack haha. We are having a hard time to extend core modules in V2 though, especially when querying the core modules with the new fields attached. But that is for another day...

**Reply by VariableVic** [reply]
Sorry to hear, that should be relatively straightforward using `defineLink` and `Query`. Have you checked out the [Extend Models resources](https://docs.medusajs.com/v2/customization/extend-models)?

**Reply by fxmb** [reply]
Yeah I checked. my problem is that i need to overwrite all endpoints. e.g. we want to add a salutation field to customers (Mr, Ms, Company). We want to extend the core module. Now I would need to adjust all customer endpoints to return this field as well, overwrite the POST (insert) endpoint to accept said saluation field etc..

**Reply by fxmb** [reply]
https://docs.medusajs.com/v2/customization/extend-models/query-linked-records

See here? I dont want to add a new custom route, I would want the existing store/customers/me endpoint to return the salutation field out of the box. I cannot rewrite every single endpoint that involves a customer....

**Reply by fxmb** [reply]
Sorry to ask here but in Discord no one is responding... you dont happen to have an example in your repo here where you add additional fields to an endpoint / Extend an existing model? Couldnt find anything when search here.
